python/ray/data/dataset_pipeline.py

```diff
@@ -652,8 +662,14 @@ def iter_datasets(self) -> Iterator[Dataset[T]]:
         if self._executed[0]:
             raise RuntimeError("Pipeline cannot be read multiple times.")
         self._executed[0] = True
-        self._optimize_stages()
-        return PipelineExecutor(self)
+        iter = None
+        if self._dataset_iter:
+            iter = itertools.chain([self._first_dataset], self._dataset_iter)
+            del self._dataset_iter
```

Comment:
We should set the iterator to `None` instead of deleting the attribute, otherwise we'd get an `AttributeError` on later accesses.
```suggestion
            del self._dataset_iter
```

Reply:
Done.